residue_genomic_position_script.py

```diff
@@ -0,0 +1,266 @@
+
+
+# Global configuration for Spark and Pandarallel.
+spark = SparkSession.builder.master('local[*]').getOrCreate()
+
+pandarallel.initialize(
+    nb_workers=psutil.cpu_count(),
+    progress_bar=True,
+)
+
+def main():
+
+    # PLIP INPUT (contain wanted data)
+    plip_json_input = (
+        # "gene_mapped_structures.json"
+        spark.read.json(args.plip_input)
+
+        .select("pdbStructureId", "chains", f.explode("compoundIds").alias("pdbCompoundId"))
+
+        .select("pdbStructureId", "pdbCompoundId", f.explode("chains").alias("chains"))
+
+        )
+
+    plip_json_input_v2 = (plip_json_input
+
+                        .withColumn("chainId", plip_json_input["chains.chainId"])
+
+                        .withColumn("geneId", plip_json_input["chains.geneId"])
+
+                        .withColumn("uniprotId", plip_json_input["chains.uniprot"])
+
+                        .drop("chains")
+
+                        )
+
+    # PLIP OUTPUT
+    plip_csv_output = (
+
+        # "output.csv"
+        spark.read.csv(args.plip_output, header=True, sep=",")
+
+        .withColumnRenamed("pdb_structure_id", "pdbStructureId")
+
+        .withColumnRenamed("compound_id", "pdbCompoundId")
+
+        .withColumnRenamed("prot_chain_id", "chainId")
+
+        )
+
+    # JOIN to have gene id (used for filter mapping file on the gene id)
+    plip_output_target_id = (
+
+        plip_json_input_v2
+
+        .join(plip_csv_output, on=["pdbStructureId", "chainId", "pdbCompoundId"])
+
+        .withColumnRenamed("interaction_type", "intType")
+
+        .withColumnRenamed("prot_residue_number", "protResNb")
+
+        .withColumnRenamed("prot_residue_type", "protResType")
+
+    )
+
+    # Target df
+    target_df = (
+
+        spark.read
+
+        .parquet("../targets")
+
+        .select("id", "genomicLocation")
+
+        .withColumn("chromosome", f.col("genomicLocation.chromosome"))
+        
+        .withColumnRenamed("id", "geneId")
+
+        .drop("genomicLocation")
+    )
+
+    plip_output_agg = (
+
+        plip_output_target_id
+        
+        .join(target_df, on='geneId')
+
+        .groupby([f.col('geneId'),
+                f.col('uniprotId'),
+                f.col("pdbStructureId").alias("pdbStructId")
+                ])
+
+        .agg(f.collect_set(f.col("pdbCompoundId")).alias("pdbCompId"),
+
+            f.collect_set(f.struct(
+                f.col('chromosome'),
+                f.col('intType'),
+                f.col('chainId'),
+                f.col('protResType'),
+                f.col('protResNb')))        
+            .alias("chr, intType, chain, resType, resNb")
+            )
+        )
+
+    # Test set
+    if args.test_set:
+
+        plip_output_agg = plip_output_agg.sample(0.001, 3)
+
+    # # Pandas Apply
+    genomic_pos_pd = plip_output_agg.toPandas()
+    genomic_pos_pd["resInfos"] = genomic_pos_pd.parallel_apply(
+        fetch_gapi_ensembl_mapping, axis=1
+        )
+
```

Comment:
There's no need to convert everything to pandas... we can select only the relevant columns, do the processing, convert back to spark dataframe and join back. We have to work out this.

Reply:
3 on 4 columns are needed for the apply so I just converted the whole Df